x-pack/plugins/apm/common/agent_configuration/runtime_types/bytes_rt.ts

```diff
@@ -6,28 +6,74 @@
 
 import * as t from 'io-ts';
 import { either } from 'fp-ts/lib/Either';
+import { i18n } from '@kbn/i18n';
 import { amountAndUnitToObject } from '../amount_and_unit';
+import { getRangeType } from './get_range_type';
 
-export const BYTE_UNITS = ['b', 'kb', 'mb'];
-
-export const bytesRt = new t.Type<string, string, unknown>(
-  'bytesRt',
-  t.string.is,
-  (input, context) => {
-    return either.chain(t.string.validate(input, context), inputAsString => {
-      const { amount, unit } = amountAndUnitToObject(inputAsString);
-      const amountAsInt = parseInt(amount, 10);
-      const isValidUnit = BYTE_UNITS.includes(unit);
-      const isValid = amountAsInt > 0 && isValidUnit;
-
-      return isValid
-        ? t.success(inputAsString)
-        : t.failure(
-            input,
-            context,
-            `Must have numeric amount and a valid unit (${BYTE_UNITS})`
-          );
-    });
-  },
-  t.identity
-);
+function toBytes(amount: number, unit: string) {
+  switch (unit) {
+    case 'kb':
+      return amount * 2 ** 10;
+    case 'mb':
+      return amount * 2 ** 20;
+    case 'b':
+    default:
+      return amount;
+  }
+}
+
+export function getBytesRt({
+  min,
+  max,
+  units
+}: {
+  min?: string;
+  max?: string;
+  units: string[];
+}) {
+  const { amount: minAmount, unit: minUnit } = min
+    ? amountAndUnitToObject(min)
+    : { amount: -Infinity, unit: 'b' };
+
+  const { amount: maxAmount, unit: maxUnit } = max
+    ? amountAndUnitToObject(max)
+    : { amount: Infinity, unit: 'mb' };
+
+  const message = i18n.translate('xpack.apm.agentConfig.bytes.errorText', {
+    defaultMessage: `{rangeType, select,
+        between {Must be between {min} and {max} with unit: {units}}
+        gt {Must be greater than {min} with unit: {units}}
+        lt {Must be less than {max} with unit: {units}}
+        other {Must be an integer with unit: {units}}
+      }`,
+    values: {
+      min,
+      max,
+      units: units.join(', '),
+      rangeType: getRangeType(minAmount, maxAmount)
+    }
+  });
+
+  return new t.Type<string, string, unknown>(
+    'bytesRt',
+    t.string.is,
+    (input, context) => {
+      return either.chain(t.string.validate(input, context), inputAsString => {
+        const { amount, unit } = amountAndUnitToObject(inputAsString);
+        const isValidUnit = units.includes(unit);
+
+        const inputAsBytes = toBytes(amount, unit);
+        const minAsBytes = toBytes(minAmount, minUnit);
+        const maxAsBytes = toBytes(maxAmount, maxUnit);
```

Comment:
I think we can simplify a lot of this by extracting some of this like:

```ts
const minAsBytes = amountAndUnitToBytes(min);
const maxAsBytes = amountAndUnitToBytes(min);
const message = getRangeTypeMessage(min, max);
```

where `amountAndUnitToBytes` looks something like:
```ts
function amountAndUnitToBytes(value?: string): number | undefined {
  const { amount, unit } = amountAndUnitToObject(value);
  if (amount != null && unit != null) {
    return toBytes(amount, unit);
  }
}
```